internal/runner/accept.go

```diff
@@ -89,63 +88,131 @@ func (r *Runner) AcceptExact(ctx context.Context, id string) error {
+	var streamCancel context.CancelFunc
+	var streamCtxLock sync.Mutex
+	var stateGen uint64
+	var err error
+
+	// We wrap this in a func() so that we use the latest client value
+	// and don't stack defers on retry.
+	defer func() {
+		if client != nil {
+			client.CloseSend()
+		}
 
-				retry = true
-				continue
-			}
+		streamCtxLock.Lock()
+		defer streamCtxLock.Unlock()
+		if streamCancel != nil {
+			// Wrap in a func() so that if we retry, we don't stack defers.
+			streamCancel()
+		}
+	}()
 
-			return err
+	// If we have a timeout, then we setup a timer for accepting.
+	var acceptTimer *time.Timer
+	var canceled int32
+	if r.acceptTimeout > 0 {
+		acceptTimer = time.AfterFunc(r.acceptTimeout, func() {
+			log.Error("runner timed out waiting for a job",
+				"timeout", r.acceptTimeout.String())
+
+			// Grab this lock before updating canceled. You don't
+			// need to have this lock to touch canceled (we use atomic ops)
+			// but it is used when the streamCancel is being reset so that
+			// we don't set it up and race with cancelation.
+			streamCtxLock.Lock()
+			defer streamCtxLock.Unlock()
+
+			// Mark that we canceled
+			atomic.StoreInt32(&canceled, 1)
+
+			// Cancel the context
+			streamCancel()
+		})
+	}
+
+RESTART_JOB_STREAM:
+	// If we're retrying, these might be non-nil and we want to do some clean-up
+	if retry {
+		log.Warn("server down before accepting a job, will reconnect")
+
+		if client != nil {
+			client.CloseSend()
+		}
+	}
+
+	// Setup a new context that we can cancel at any time to close the stream.
+	// We use this for timeouts.
+	//
+	// Note: we disable the lostcancel linter for streamCancel because
+	// golangci-lint is not detecting that we have the defer above the
+	// label as well as the retry block above.
+	streamCtxLock.Lock()
+	if streamCancel != nil {
+		streamCancel()
+	}
+	if atomic.LoadInt32(&canceled) > 0 {
+		streamCtxLock.Unlock()
+		return ErrTimeout
+	}
+	streamCtx, streamCancel = context.WithCancel(r.runningCtx)
+	streamCtxLock.Unlock()
+
+	// Since this is a disconnect, we have to wait for our
+	// RunnerConfig stream to re-establish. We wait for the config
+	// generation to increment.
+	if retry {
+		if r.waitStateGreater(&r.stateConfig, stateGen) {
+			return status.Error(codes.Internal, "early exit while waiting for reconnect")
 		}
+	}
+
+	// Get our configuration state value. We use this so that we can detect
+	// when we've reconnected during failures.
+	stateGen = r.readState(&r.stateConfig)
 
-		break
+	// Open a new job stream. This retries on connection errors. Note that
+	// this retry loop doesn't respect the accept timeout because gRPC has no way
+	// to time out of a "WaitForReady" RPC call (it ignores context cancellation,
+	// too). TODO: do a manual backoff with WaitForReady(false) so we can
+	// weave in accept timeout.
+	// NOTE: we purposely do NOT use ctx above since if the context is
+	// cancelled we want to continue reporting errors.
```

Comment:
I think this comment should come around line 181 instead? Is that the situation it's referring to?